src/main/scala/org/wikimedia/analytics/hdfstools/HdfsRsyncExec.scala

```diff
@@ -257,91 +237,83 @@ class HdfsRsyncExec(config: HdfsRsyncConfig) {
      * @return Some path if dst was defined (can be a non-existent file in case of dryrun),
      *         None if no dst was provided (logs only).
      */
-    private def processSrcAndDst(
+    private def processFile(
         src: FileStatus,
-        dst: Option[Path],
-        foundTarget: Option[FileStatus],
-        copy: Boolean
-    ): Option[Path] = {
+        target: Option[Path],
+        foundTarget: Option[FileStatus]
+    ): Unit = {
         // If no destination is provided, just log the file to copy
-        if (config.dst.isEmpty) {
-            if (copy) log.info(s"COPY_FILE [no-dst] - ${src.getPath}")
-            else log.info(s"CREATE_FOLDER [no-dst] - ${src.getPath}")
-            None
+        if (target.isEmpty) {
+            log.info(s"COPY_FILE [no-dst] - ${src.getPath}")
         } else {
-            val target = new Path(s"${dst.get.toString}/${src.getPath.getName}")
             val isNew = foundTarget.isEmpty
             val targetToUpdate = {
                 if (isNew) {
-                    createOrCopyNew(src, target, copy)
+                    copyNewFile(src, target.get)
                 } else if (config.ignoreTimes || // Force copy/update
                     src.getLen != foundTarget.get.getLen || // Size is different
                     (!config.sizeOnly &&                    // Use modificationTime
                         approxDiffTimestamps(src, foundTarget.get))) {
-                    copyUpdate(src, target, copy)
+                    updateFile(src, target.get)
                 } else {
-                    skip(src, target, copy)
+                    skipFile(src, target.get)
                 }
             }
-            preserveModificationTimesAsNeeded(src, targetToUpdate)
-            updatePermsAsNeeded(src, targetToUpdate, isDir = src.isDirectory, isNew)
-            Some(target)
+            if (config.preserveTimes) {
+                preserveModificationTimes(src, targetToUpdate)
+            }
+            // Update perms if needed
+            if (config.preservePerms || (isNew && (isNew && config.chmodFiles.isDefined))) {
+                updateFilePerms(src, targetToUpdate, config.chmodFiles)
+            }
         }
     }
 
     /**
-     * Function listing files and folders of a src folder or glob (first call before recursion),
-     * and building a Map[filename, FileStatus] from it.
+     * Function processing a single folder with a target and foundTarget.
+     * The target folder is created if it doesn't exist, and its permissions are
+     * updated depending on configuration.
      *
-     * @param srcPath the src path to scan
-     * @param isBasePath whether to use a glob if true or a folder scan if false
-     * @return the Map[filename, FileStatus] of the folder/glob content
+     * @param target the target where to create the new folderif necessary (if any)
+     * @param foundTarget the found target in dst (same filename as src) if any
```

Comment:
I don't understand the difference between target and foundTarget, aren't target's always in dst?